Doc/library/signal.rst

```diff
@@ -503,3 +503,32 @@ be sent, and the handler raises an exception. ::
 
    signal.alarm(0)          # Disable the alarm
 
+Note on SIGPIPE
+---------------
+
+Piping output of your program to tools like :manpage:`head(1)` will
+cause a SIGPIPE signal to be sent to your process when the receiver
+of its standard output closes early.  This results in an exception
+like :code:`BrokenPipeError: [Errno 32] Broken pipe`.  To handle this
+case, wrap your entry point to catch this exception as follows::
+
+    import os
+    import sys
+
+    def main():
+        try:
+            for x in range(10000):
+                print("y")
+            # flush output here to force SIGPIPE to be triggered
+            # while inside this try block.
+            sys.stdout.flush()
+        except BrokenPipeError:
+            os._exit(1)  # Python exits with error code 1 on EPIPE
```

Comment:
If you use os.exit() the following output happens:

```
22:56:46 ~ % cat t2.py 
import os
import sys

def main():
    try:
        for x in range(10000):
            print("y")
        # flush output here to force SIGPIPE to be triggered
        # while inside this try block.
        sys.stdout.flush()
    except BrokenPipeError:
        sys.exit(1)  # Python exits with error code 1 on EPIPE

if __name__ == '__main__':
    main()

22:56:50 ~ % python3 t2.py | head -1
y
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
22:56:52 ~ % 

```

Since I can't seem to find a way for python NOT to flush on sys.exit(), this code works, but is very ugly, however it's safer since you are correct about os._exit() skipping shutdown handlers....

This works, so maybe we should amend the code to be like this (removing the unneeded comments and writes to stderr):
```
22:59:21 ~ % cat t.py 
import os
import sys

def main():
    try:
        for x in range(10000):
            print("y")
        # flush output here to force SIGPIPE to be triggered
        # while inside this try block.
        sys.stdout.flush()
    except BrokenPipeError:
        sys.stderr.write("BPE 1\n")
        try:
            sys.stdout.close()
            #sys.stdout.flush()
            pass
        except BrokenPipeError:
            sys.stderr.write("BPE 2\n")
            pass
        #os._exit(1)  # Python exits with error code 1 on EPIPE
        sys.exit(1)  # Python exits with error code 1 on EPIPE
    finally:
        sys.stdout.close()


if __name__ == '__main__':
    main()

22:59:26 ~ % python3 t.py | head -1 
y
BPE 1
BPE 2
22:59:47 ~ % 

```

The problem here is that python doesn't seem to have a method to "fpurge(3)" a stream which would work better than double closing it.

Reply:
This seems to work... I wish there was a way to "fpurge(3)" a python file object instead.

```
import os
import sys

def main():
    try:
        for x in range(10000):
            print("y")
        # flush output here to force SIGPIPE to be triggered
        # while inside this try block.
        sys.stdout.flush()
    except BrokenPipeError:
        devnull = os.open(os.devnull, os.O_WRONLY)
        os.dup2(devnull, sys.stdout.fileno())
        sys.exit(1)  # Python exits with error code 1 on EPIPE

if __name__ == '__main__':
    main()

```

Reply:
Also python doesn't exit with the correct status code when it catches sigpipe, it returns with exit code 1.  (I tested this with Lukasz)

Reply:
I'm fine with the latest iteration. We just need to comment on *why* we're duping stdout to devnull. As simple as "Python flushes standard streams on exit; redirect remaining output to devnull to avoid another BrokenPipeError at shutdown".